### PR TITLE
Report that a project only has one TFM

### LINKED ISSUE #739: Multi-proc build of a .sln file is 'The process cannot access the file because it is being used by another process.'
In the CLI repo, we have a Test.sln file which we `dotnet build`.  However, we are seeing random failures that look like the following:

```
05:25:35.797     1>CSC : error CS2012: Cannot open 'D:\j\workspace\debug_windows---54dcb9ec\test\Msbuild.Tests.Utilities\obj\Debug\netcoreapp1.0\Msbuild.Tests.Utilities.dll' for writing -- 'The process cannot access the file 'D:\j\workspace\debug_windows---54dcb9ec\test\Msbuild.Tests.Utilities\obj\Debug\netcoreapp1.0\Msbuild.Tests.Utilities.dll' because it is being used by another process.' [D:\j\workspace\debug_windows---54dcb9ec\test\Msbuild.Tests.Utilities\Msbuild.Tests.Utilities.csproj] [D:\j\workspace\debug_windows---54dcb9ec\build.proj]
```

See https://ci.dot.net/job/dotnet_cli/job/rel_1.0.0/job/debug_windows_nt_x86_prtest/810/console

To repro:

1. On Windows x64 machine, clone the CLI repo and checkout https://github.com/dotnet/cli/pull/5465/commits/f6b0196b0f8a7342d4b5fe5709568af9e3ede665.  
2. Then remove the line https://github.com/dotnet/cli/pull/5465/commits/f6b0196b0f8a7342d4b5fe5709568af9e3ede665#diff-70eb062c7648e65aa7aff25e924d014bR84 which is working around this problem by turning off multi-proc build.
3. build.cmd

An easier repro once you have built one time is:

0. `F:\cli>git clean test -xdf`
1. `F:\cli>.dotnet_stage0\x64\dotnet.exe restore test\Microsoft.DotNet.Cli.Tests.sln`
2. `F:\cli>.dotnet_stage0\x64\dotnet.exe build test\Microsoft.DotNet.Cli.Tests.sln`

```
F:\cli>.dotnet_stage0\x64\dotnet.exe build test\Microsoft.DotNet.Cli.Tests.sln
Microsoft (R) Build Engine version 15.1.523.56541
Copyright (C) Microsoft Corporation. All rights reserved.

F:\cli\.dotnet_stage0\x64\sdk\2.0.0-alpha-004616\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.DefaultItems.targets(101,5): warning : A PackageReference for 'NETStandard.Library' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [F:\cli\test\Microsoft.DotNet.Tools.Tests.Utilities\Microsoft.DotNet.Tools.Tests.Utilities.csproj]
F:\cli\.dotnet_stage0\x64\sdk\2.0.0-alpha-004616\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.DefaultItems.targets(101,5): warning : A PackageReference for 'NETStandard.Library' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [F:\cli\test\Microsoft.DotNet.Tools.Tests.Utilities\Microsoft.DotNet.Tools.Tests.Utilities.csproj]
  ArgumentsReflector -> F:\cli\test\ArgumentsReflector\bin\Debug\netcoreapp2.0\ArgumentsReflector.dll
  Microsoft.DotNet.Cli.Utils -> F:\cli\src\Microsoft.DotNet.Cli.Utils\bin\Debug\netstandard1.5\Microsoft.DotNet.Cli.Utils.dll
  Microsoft.DotNet.TestFramework -> F:\cli\src\Microsoft.DotNet.TestFramework\bin\Debug\netstandard1.5\Microsoft.DotNet.TestFramework.dll
  Microsoft.DotNet.InternalAbstractions -> F:\cli\src\Microsoft.DotNet.InternalAbstractions\bin\Debug\netstandard1.3\Microsoft.DotNet.InternalAbstractions.dll
CSC : error CS2012: Cannot open 'F:\cli\test\Microsoft.DotNet.Tools.Tests.Utilities\obj\Debug\netstandard1.5\Microsoft.DotNet.Tools.Tests.Utilities.dll' for writing -- 'The process cannot access the file 'F:\cli\test\Microsoft.DotNet.Tools.Tests.Utilities\obj\Debug\netstandard1.5\Microsoft.DotNet.Tools.Tests.Utilities.dll' because it is being used by another process.' [F:\cli\test\Microsoft.DotNet.Tools.Tests.Utilities\Microsoft.DotNet.Tools.Tests.Utilities.csproj]
```

### PR DESCRIPTION
**Pull request by rainersigwald**
Partially fixes dotnet/sdk#739.

The first iteration of cross-targeting support code unconditionally
queried each ProjectReference for the best TFM to build against, and
then explicitly specified that TFM when building the reference. This
caused a race condition when building a set of projects that had a
single-TFM project and another project that had a reference to it. The
entry point (probably .sln) build would build the referenced project
with no TF specified, and then the referencing project would build it
with an explicit TF specified. These two builds appeared different to
the MSBuild engine (because they had different sets of global
properties) and so were both executed, resulting in a race condition.

The fix is in two parts:
* Allow a project to report back through GetTargetFrameworkProperties
that it only has one TF to build.
* When a project reports that it has only one TF to build, issue its
build request without specifying any TF. (https://github.com/Microsoft/msbuild/pull/1667)

This commit is the former.

### CODE REVIEW

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets [hunk 1]
```diff
@@ -55,13 +55,16 @@ Copyright (c) .NET Foundation. All rights reserved.
     with the referencing project's target framework.
   ============================================================
    -->
-  <Target Name="GetTargetFrameworkProperties" Returns="TargetFramework=$(NearestTargetFramework)">
+  <Target Name="GetTargetFrameworkProperties" Returns="TargetFramework=$(NearestTargetFramework);ProjectHasSingleTargetFramework=$(_HasSingleTargetFramework)">
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: Can we use metadata to indicate ProjectHasSingleTargetFramework? Seems like it would simplify the consuming side.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets [hunk 2]
```diff
@@ -55,13 +55,16 @@ Copyright (c) .NET Foundation. All rights reserved.
...
 
     <PropertyGroup>
       <!-- If a ReferringTargetFramework was not specified, and we only have one TargetFramework, then don't try to check compatibility -->
       <_SkipNearestTargetFrameworkResolution Condition="'$(TargetFramework)' != '' and '$(ReferringTargetFramework)' == ''">true</_SkipNearestTargetFrameworkResolution>
       <NearestTargetFramework Condition="'$(_SkipNearestTargetFrameworkResolution)' == 'true'">$(TargetFramework)</NearestTargetFramework>
       
+      <_HasSingleTargetFramework Condition="'$(TargetFramework)' != '' and '$(TargetFrameworks)' == ''">true</_HasSingleTargetFramework>
+      <_HasSingleTargetFramework Condition="'$(_HasSingleTargetFramework)' == ''">false</_HasSingleTargetFramework>
+
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: I believe everywhere else we consider TargetFramework to win over TargetFrameworks, so this might be better as TargetFramework != ''

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets (hunk 1 above)
Comment (rainersigwald) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: Indeed it would, but it requires returning an item rather than a property which seemed more invasive.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets (hunk 2 above)
Comment (rainersigwald) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: I don't think I understand. Can you rephrase?

Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: If I have a project that sets both TargetFramework and TargetFrameworks, TargetFramework will win (we won't go to cross-targeting targets), but this won't report it as such.

**Comment by rainersigwald** [comment]
Alternative with an item + metadata instead:

```diff
diff --git "a/c:\\Program Files\\dotnet - Copy\\sdk\\1.0.0-rc4-004689/Microsoft.Common.CurrentVersion.targets" "b/C:\\Program Files\\dotnet\\sdk\\1.0.0-rc4-004689/Microsoft.Common.CurrentVersion.targets"
index da772b0..96e8fb4 100644
--- "a/c:\\Program Files\\dotnet - Copy\\sdk\\1.0.0-rc4-004689/Microsoft.Common.CurrentVersion.targets"	
+++ "b/C:\\Program Files\\dotnet\\sdk\\1.0.0-rc4-004689/Microsoft.Common.CurrentVersion.targets"	
@@ -1524,13 +1524,20 @@ Copyright (C) Microsoft Corporation. All rights reserved.
         RemoveProperties="%(_MSBuildProjectReferenceExistent.GlobalPropertiesToRemove);TargetFramework"
         Condition="'%(_MSBuildProjectReferenceExistent.Extension)' != '.vcxproj'">
 
-      <Output TaskParameter="TargetOutputs" PropertyName="_ProjectReferenceTargetFrameworkProperties" />
+      <Output TaskParameter="TargetOutputs" ItemName="_ProjectReferenceTargetFrameworkProperties" />
     </MSBuild>
 
+    <Error Condition="@(_ProjectReferenceTargetFrameworkProperties->Count()) != 1" Text="Assertion failed: @(_ProjectReferenceTargetFrameworkProperties->Count()) @(_ProjectReferenceTargetFrameworkProperties)" />
+
     <ItemGroup>
       <_MSBuildProjectReferenceExistent Condition="'%(_MSBuildProjectReferenceExistent.Identity)' == '%(Identity)'">
         <SetTargetFramework>$(_ProjectReferenceTargetFrameworkProperties)</SetTargetFramework>
+
+        <UndefineProperties Condition="'@(_ProjectReferenceTargetFrameworkProperties->'%(HasSingleTargetFramework)')' == 'true'">%(_MSBuildProjectReferenceExistent.UndefineProperties);TargetFramework</UndefineProperties>
       </_MSBuildProjectReferenceExistent>
+
+      <!-- Clear the list so the next iteration through the loop has a single item again. -->
+      <_ProjectReferenceTargetFrameworkProperties Remove="@(_ProjectReferenceTargetFrameworkProperties)" />
     </ItemGroup>
 
     <PropertyGroup>
diff --git "a/c:\\Program Files\\dotnet - Copy\\sdk\\1.0.0-rc4-004689/Sdks/Microsoft.NET.Sdk/build/Microsoft.NET.Sdk.Common.targets" "b/C:\\Program Files\\dotnet\\sdk\\1.0.0-rc4-004689/Sdks/Microsoft.NET.Sdk/build/Microsoft.NET.Sdk.Common.targets"
index c1c71c2..ae43aa0 100644
--- "a/c:\\Program Files\\dotnet - Copy\\sdk\\1.0.0-rc4-004689/Sdks/Microsoft.NET.Sdk/build/Microsoft.NET.Sdk.Common.targets"	
+++ "b/C:\\Program Files\\dotnet\\sdk\\1.0.0-rc4-004689/Sdks/Microsoft.NET.Sdk/build/Microsoft.NET.Sdk.Common.targets"	
@@ -55,7 +55,7 @@ Copyright (c) .NET Foundation. All rights reserved.
     with the referencing project's target framework.
   ============================================================
    -->
-  <Target Name="GetTargetFrameworkProperties" Returns="TargetFramework=$(NearestTargetFramework)">
+  <Target Name="GetTargetFrameworkProperties" Returns="@(ProjectWithTargetFrameworkProperties)">
 
     <PropertyGroup>
       <!-- If a ReferringTargetFramework was not specified, and we only have one TargetFramework, then don't try to check compatibility -->
@@ -72,6 +72,14 @@ Copyright (c) .NET Foundation. All rights reserved.
                                Condition="'$(_SkipNearestTargetFrameworkResolution)' != 'true'">
       <Output PropertyName="NearestTargetFramework" TaskParameter="NearestTargetFramework" />
     </GetNearestTargetFramework>
+
+    <ItemGroup>
+      <ProjectWithTargetFrameworkProperties Include="$(MSBuildProjectFullPath)">
+        <TargetFramework>$(NearestTargetFramework)</TargetFramework>
+        <HasSingleTargetFramework Condition="'$(TargetFrameworks)' != ''">false</HasSingleTargetFramework>
+        <HasSingleTargetFramework Condition="'$(TargetFramework)' != '' and '$(TargetFrameworks)' == ''">true</HasSingleTargetFramework>
+      </ProjectWithTargetFrameworkProperties>
+    </ItemGroup>
   </Target>
   
 </Project>
```

I don't think I prefer it. Have to jump through hoops to operate on a single result at a time, where that's implicit with the property. Not that the property way in the PR is clean or nice . . .

**Comment by nguerrera** [comment]
Hmm that is a lot messier than I imagined. I trust your judgement on this.

**Comment by srivatsn** [comment]
@natidea can you test this change with whatever repros we've got?

**Comment by srivatsn** [comment]
You'll need this change as well - https://github.com/Microsoft/msbuild/pull/1667

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets [hunk 3]
```diff
@@ -55,13 +55,16 @@ Copyright (c) .NET Foundation. All rights reserved.
...
    -->
-  <Target Name="GetTargetFrameworkProperties" Returns="TargetFramework=$(NearestTargetFramework)">
+  <Target Name="GetTargetFrameworkProperties" Returns="TargetFramework=$(NearestTargetFramework);ProjectHasSingleTargetFramework=$(_HasSingleTargetFramework)">
 
     <PropertyGroup>
       <!-- If a ReferringTargetFramework was not specified, and we only have one TargetFramework, then don't try to check compatibility -->
       <_SkipNearestTargetFrameworkResolution Condition="'$(TargetFramework)' != '' and '$(ReferringTargetFramework)' == ''">true</_SkipNearestTargetFrameworkResolution>
       <NearestTargetFramework Condition="'$(_SkipNearestTargetFrameworkResolution)' == 'true'">$(TargetFramework)</NearestTargetFramework>
       
+      <_HasSingleTargetFramework Condition="'$(TargetFramework)' != ''">true</_HasSingleTargetFramework>
```
Comment (srivatsn) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: If a SDK-based project doesn't define a TargetFramework, do we issue an error today? I'm thinking if someone takes an existing project where TFI, TFV are set and makes it use the SDK and doesn't set TargetFramework, will we now think of them as not-single-targeting projects?

Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: Good point. I think we need this to be `'$(IsCrossTargetingBuild)' != 'true'`. Whatever condition causes us not to use the x-targeting is the same as what should get no additional properties.

Comment (rainersigwald) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: Makes perfect sense. Thanks.

**Comment by rainersigwald** [comment]
@natidea MSBuild package 545 should have the common-targets side of this baked in.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets [hunk 4]
```diff
@@ -29,7 +29,7 @@ Copyright (c) .NET Foundation. All rights reserved.
     <Version Condition=" '$(Version)' == '' ">$(VersionPrefix)</Version>
     <Authors Condition=" '$(Authors)'=='' ">$(AssemblyName)</Authors>
     <Company Condition=" '$(Company)'=='' ">$(Authors)</Company>
-    <AssemblyTitle Condition="'$(AssemblyTitle)' == ''">$(AssemblyName)</AssemblyTitle>
+-    <AssemblyTitle Condition="'$(AssemblyTitle)' == ''">$(AssemblyName)</AssemblyTitle>
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: Looks like there are some whitespace and extra leading character issues

Comment (rainersigwald) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: That's what I get for trusting the GitHub web editor.

Comment (rainersigwald) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: Fixed.

**Review by nguerrera — APPROVED** [review]

**Comment by nguerrera** [comment]
**Customer scenario**: Parallel solution builds with single-targeted sdk-based projects fail intermittently (but frequently)

**Bugs this fixes:**: #739

**Workarounds, if any**: None

**Risk**: Low

**Performance impact**: Low

**Is this a regression from a previous update?**: No

**Root cause analysis:**: Complicated interaction between cross-targeting P2P protocol, sln builds, and /m > 1.

**How was the bug found?**: Customer and partner reported

**Comment by nguerrera** [comment]
@MattGertz for approval

**Comment by MattGertz** [comment]
@srivatsn Feel free to add this to your VSO bug now if you are comfortable.  It'll be another 40 minutes before they get to us...

**Comment by srivatsn** [comment]
Added

**Comment by natidea** [comment]
Shiproom Update: Waiting for Weekend validation

**Review by natidea — APPROVED** [review]
After applying both patches, I no longer see the race through multiple builds of both CLI build and DependencyInjection.sln :shipit:
